### ISSUE TITLE
TcpClient very slow connecting to "localhost"

### ISSUE THREAD

**Issue by inmemorynet** [dotnet/corefx#41588]
# Issue Title

TcpClient very slow connecting to "localhost"

# General

We are seeing a major performance regression when connecting using "localhost" . This use to take less than 1 millisecond, but now takes over 2 seconds.  The following snipped replicates the behaviour.

TcpClient client = new TcpClient();
client.Connect ("localhost",5060);

Using 127.0.0.1 as a workaround does not have this problem.
The code in Dot Net Framework 4,4.5,4.7,4.8, and Dot Net Core 2.1 does not exhibit this delay.

Connecting to IIS on port 80, does not exhibiting this delay.

Tested against latest version of Windows 10, and Visual Studio 16,3,2, and Dot Net Core 3.0 RTM

We are experiencing this on our InMemory.Net Database server. This can be downloaded from https://www.inmemory.net

**Comment by Tornhoof** [comment]
Try using 127.0.0.1 directly. I've seen that problem,  it tried to connect to ::1 (ipv6) first, because localhost resolved to [::1,127.0.0.1].

**Comment by stephentoub** [comment]
> The code in Dot Net Framework 4,4.5,4.7,4.8, and Dot Net Core 2.1 does not exhibit this delay.

Are you saying this difference occurs on the same application build on the same machine on the same OS on the same network, etc., and all you change is whether you run with .NET Core 2.1 vs .NET Core 3.0?

My experience matches that of @Tornhoof: there's likely actually a difference outside of the version of .NET that's causing DNS resolution to return an IPv6 address as the first result instead of an IPv4.

**Comment by inmemorynet** [comment]
Yes, I wanted to confirm, that the behaviour is happening in the same machine / os , and all I change is .Net Core 2.1 to .Net Core 3.0

**Comment by Tornhoof** [comment]
A bit additional info (as I tried to figure this out before):
- For me this is reproducable with 2.1+ on current Win10 Desktops (and  Server) I think roughly since the 1903 update.
- I noticed it because some integration tests with TcpClients suddenly took close to two seconds longer (and with 500+ tests this was rather noticeable).
- I don't think it's .NET Core related, as I can reproduce it with the googlechromedriver for selenium too. As soon as you add verbose logging to the google chrome driver, it outputs the resolved hostnames and that was [::1,127.0.0.1] for localhost and it tried to connect to ipv6 first.
- I don't know why e.g. HttpClient is not affected.

You can try it yourself by something like, if the output is ::1 first, then it's the problem I described.
```csharp
    static async Task Main(string[] args)
    {

        foreach (IPAddress address in await Dns.GetHostAddressesAsync("localhost"))
        {
            Console.WriteLine(address);
        }
    }
```

For e.g. https://dotnetfiddle.net/4LANAN it's the other way round. (ipv4 first)

**Comment by davidsh** [comment]
FWIW, the implementation of Dns.GetHostAddressesAsync() changed from .NET Core 2.1 to .NET Core 3.0 on Windows.

**Comment by stephentoub** [comment]
> the implementation of Dns.GetHostAddressesAsync() changed from .NET Core 2.1 to .NET Core 3.0 on Windows

That change was in 2.1.
https://github.com/dotnet/corefx/pull/26850

**Comment by hpetriffer** [comment]
I have an identical problem with an application. If I use localhost it is very slow. With 127.0.0.1 it works very fast.
I use also .net core 3.0

**Comment by wfurt** [comment]
you can always run Wireshark and filter for any traffic to to 5060. If you see failing IPv6 attempts @inmemorynet  that would confirm the suspicion. (80 may work if configured for both address families) You can also specify address family explicitly for TcpClient().

**Comment by hpetriffer** [comment]
First I get an IPv6 attempt...

**Comment by scalablecory** [comment]
#41778 will help here.

**Comment by wfurt** [comment]
note that normally the IPv6 connect should fail fast with RST unless blocked by firewall. 
That is something to try and check with Wireshark.

**Comment by karelz** [comment]
Duplicate of #41778
